### ISSUE TITLE
action_delay is not accurate 

### ISSUE THREAD

**Issue by thuang6** [intel/workload-collocation-agent#170]
In my config, I set action_delay to 1, every 20 cycle, I do some metrics aggregation and detection.
the period is not accurate, I add debug print every time allocate function is invoked

it takes about 12s when I count to 19

debug output:

2019-06-19 08:00:44,190 INFO     {MainThread} [prm.allocator] counter: 0
2019-06-19 08:00:44,518 INFO     {MainThread} [prm.allocator] counter: 1
2019-06-19 08:00:45,434 INFO     {MainThread} [prm.allocator] counter: 2
2019-06-19 08:00:45,789 INFO     {MainThread} [prm.allocator] counter: 3
2019-06-19 08:00:46,695 INFO     {MainThread} [prm.allocator] counter: 4
2019-06-19 08:00:47,049 INFO     {MainThread} [prm.allocator] counter: 5
2019-06-19 08:00:47,952 INFO     {MainThread} [prm.allocator] counter: 6
2019-06-19 08:00:48,414 INFO     {MainThread} [prm.allocator] counter: 7
2019-06-19 08:00:49,222 INFO     {MainThread} [prm.allocator] counter: 8
2019-06-19 08:00:49,688 INFO     {MainThread} [prm.allocator] counter: 9
2019-06-19 08:00:50,467 INFO     {MainThread} [prm.allocator] counter: 10
2019-06-19 08:00:50,945 INFO     {MainThread} [prm.allocator] counter: 11
2019-06-19 08:00:51,690 INFO     {MainThread} [prm.allocator] counter: 12
2019-06-19 08:00:52,195 INFO     {MainThread} [prm.allocator] counter: 13
2019-06-19 08:00:52,944 INFO     {MainThread} [prm.allocator] counter: 14
2019-06-19 08:00:53,440 INFO     {MainThread} [prm.allocator] counter: 15
2019-06-19 08:00:54,169 INFO     {MainThread} [prm.allocator] counter: 16
2019-06-19 08:00:54,705 INFO     {MainThread} [prm.allocator] counter: 17
2019-06-19 08:00:55,456 INFO     {MainThread} [prm.allocator] counter: 18
2019-06-19 08:00:55,950 INFO     {MainThread} [prm.allocator] counter: 19
2019-06-19 08:00:55,959 INFO     {MainThread} [prm.allocator] pulled model thresholds={'cassandra.default': {'10.0': {'tdp_threshold': {}, 'metrics_threshold': []}}, 'redis.default': {'1.0': {'tdp_threshold': {}, 'metrics_threshold': []}}, 'twemcache.big': {'4.0': {'tdp_threshold': {}, 'metrics_threshold': []}}, 'twemcache.small': {'2.0': {'tdp_threshold': {}, 'metrics_threshold': []}}}
2019-06-19 08:00:55,959 INFO     {MainThread} [prm.allocator] cycle: 0
2019-06-19 08:00:56,706 INFO     {MainThread} [prm.allocator] counter: 0
2019-06-19 08:00:57,242 INFO     {MainThread} [prm.allocator] counter: 1
2019-06-19 08:00:57,952 INFO     {MainThread} [prm.allocator] counter: 2
2019-06-19 08:00:58,465 INFO     {MainThread} [prm.allocator] counter: 3
2019-06-19 08:00:59,202 INFO     {MainThread} [prm.allocator] counter: 4
2019-06-19 08:00:59,719 INFO     {MainThread} [prm.allocator] counter: 5
2019-06-19 08:01:00,438 INFO     {MainThread} [prm.allocator] counter: 6
2019-06-19 08:01:00,947 INFO     {MainThread} [prm.allocator] counter: 7
2019-06-19 08:01:01,698 INFO     {MainThread} [prm.allocator] counter: 8
2019-06-19 08:01:02,241 INFO     {MainThread} [prm.allocator] counter: 9
2019-06-19 08:01:02,941 INFO     {MainThread} [prm.allocator] counter: 10
2019-06-19 08:01:03,484 INFO     {MainThread} [prm.allocator] counter: 11
2019-06-19 08:01:04,191 INFO     {MainThread} [prm.allocator] counter: 12
2019-06-19 08:01:04,714 INFO     {MainThread} [prm.allocator] counter: 13
2019-06-19 08:01:05,425 INFO     {MainThread} [prm.allocator] counter: 14
2019-06-19 08:01:05,966 INFO     {MainThread} [prm.allocator] counter: 15
2019-06-19 08:01:06,655 INFO     {MainThread} [prm.allocator] counter: 16
2019-06-19 08:01:07,214 INFO     {MainThread} [prm.allocator] counter: 17
2019-06-19 08:01:07,903 INFO     {MainThread} [prm.allocator] counter: 18
2019-06-19 08:01:08,453 INFO     {MainThread} [prm.allocator] counter: 19
2019-06-19 08:01:08,455 INFO     {MainThread} [prm.allocator] cycle: 1
2019-06-19 08:01:09,230 INFO     {MainThread} [prm.allocator] counter: 0
2019-06-19 08:01:09,716 INFO     {MainThread} [prm.allocator] counter: 1
2019-06-19 08:01:10,474 INFO     {MainThread} [prm.allocator] counter: 2
2019-06-19 08:01:10,953 INFO     {MainThread} [prm.allocator] counter: 3
2019-06-19 08:01:11,732 INFO     {MainThread} [prm.allocator] counter: 4
2019-06-19 08:01:12,172 INFO     {MainThread} [prm.allocator] counter: 5
2019-06-19 08:01:15,993 INFO     {MainThread} [prm.allocator] counter: 6
2019-06-19 08:01:16,206 INFO     {MainThread} [prm.allocator] counter: 7
2019-06-19 08:01:17,220 INFO     {MainThread} [prm.allocator] counter: 8
2019-06-19 08:01:17,438 INFO     {MainThread} [prm.allocator] counter: 9
2019-06-19 08:01:18,467 INFO     {MainThread} [prm.allocator] counter: 10
2019-06-19 08:01:18,720 INFO     {MainThread} [prm.allocator] counter: 11
2019-06-19 08:01:19,714 INFO     {MainThread} [prm.allocator] counter: 12
2019-06-19 08:01:19,954 INFO     {MainThread} [prm.allocator] counter: 13
2019-06-19 08:01:20,962 INFO     {MainThread} [prm.allocator] counter: 14
2019-06-19 08:01:21,170 INFO     {MainThread} [prm.allocator] counter: 15
2019-06-19 08:01:22,204 INFO     {MainThread} [prm.allocator] counter: 16
2019-06-19 08:01:22,426 INFO     {MainThread} [prm.allocator] counter: 17
2019-06-19 08:01:23,462 INFO     {MainThread} [prm.allocator] counter: 18
2019-06-19 08:01:23,691 INFO     {MainThread} [prm.allocator] counter: 19

 ```
runner: !AllocationRunner
  node: !MesosNode
    mesos_agent_endpoint: "http://127.0.0.1:5051"
  action_delay: &action_delay 1.
  metrics_storage: !LogStorage
    output_filename: '/tmp/metrics.prom'
    overwrite: true
  anomalies_storage: !LogStorage
    output_filename: '/tmp/anomalies.prom'
    overwrite: true
  allocations_storage: !LogStorage
    output_filename: '/tmp/allocations.prom'
    overwrite: true
  allocator: !ResourceAllocator
    database: !ModelDatabase
      db_type: etcd # 1) local 2)zookeeper 3)etcd
      directory: ~     # required for local
      host: "100.64.176.27:2379"     # required for zookeeper and etcd
      namespace: ~     # for zookeeper, if none, using default model_distribution
      timeout: 5.0     # for etcd, default 5.0 seconds
      #ssl: !SSL
       # server_verify: False
       # client_cert_path: "/home/ssg/ssl/keystore.pem"
       # client_key_path: "/home/ssg/ssl/keystore.pem"
    action_delay: *action_delay
    agg_period: 20.
    model_pull_cycle: 10.
    metric_file: "/tmp/prm/metric.csv"
    enable_control: False
    exclusive_cat: False
  rdt_enabled: True
  extra_labels:
    env_uniq_id: "15"
    own_ip: "100.64.176.15"
```

**Comment by thuang6** [comment]
@ppalucki  please help to confirm if the issue is by design, if so, I will change document in PRM to avoid use some accurate time in some description

**Comment by ppalucki** [comment]
@thuang6 you are right, we've found a bug today - 

Residual time for sleeping is set dynamically depending on how much collecting metrics/external callback lasted during last iteration - this mechanism tries to amortize time spend on other things to be precise as possible to scheduled next callaback

But there was a bug introduced (details here #175), during kubernetes/allocation refactoring (two months ago and because those metrics were overlapping so we missed that in integrations tests)

fixed in #175 both master/1.0.x branch and in release 1.0.1

**Comment by thuang6** [comment]
Verified and thanks for quick fix
